In [1]:
library('tidyverse')

── Attaching packages ─────────────────────────────────────── tidyverse 1.2.1 ──

✔ ggplot2 3.2.0     ✔ purrr   0.3.2
✔ tibble  2.1.3     ✔ dplyr   0.8.3
✔ tidyr   0.8.3     ✔ stringr 1.4.0
✔ readr   1.3.1     ✔ forcats 0.4.0

── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()



In [2]:
df.marathon = read.csv('../data/marathon.csv')
df.marathon %>% head

year name                       country       time       race.time MF   
1 2019 Lawrence Cherono           Kenya         2H 7M 57S  2.132500  Men's
2 2018 Yuki Kawauchi              Japan         2H 15M 58S 2.266111  Men's
3 2017 Geoffrey Kirui             Kenya         2H 9M 37S  2.160278  Men's
4 2016 Lemi Berhanu               Ethiopia      2H 12M 45S 2.212500  Men's
5 2015 Lelisa Desisa              Ethiopia      2H 9M 17S  2.154722  Men's
6 2014 Mebrahtom "Meb" Keflezighi United States 2H 8M 37S  2.143611  Men's
  country.simple
1 Kenya         
2 Japan         
3 Kenya         
4 Ethiopia      
5 Ethiopia      
6 United States

# Filter

- the filter operation does exactly what you expect given its name.

In [3]:
# e.g.

# iris %>% filter(Species == 'setosa', Petal.Width > 0.2)

- The syntax for the `equals` operator in R is `==`.
- Greater than (`>` or `>=`) and less than (`<` or `<=`) are intuitive.
- You can fitler on multiple columns by adding arguments to the filter operation.

Filter the marathon data to just the runners from Kenya.

In [3]:
df.marathon %>% 
    filter(country == "Kenya")

year name                      country time       race.time MF     
1  2019 Lawrence Cherono          Kenya   2H 7M 57S  2.132500  Men's  
2  2017 Geoffrey Kirui            Kenya   2H 9M 37S  2.160278  Men's  
3  2012 Wesley Korir              Kenya   2H 12M 40S 2.211111  Men's  
4  2011 Geoffrey Mutai            Kenya   2H 3M 2S   2.050556  Men's  
5  2010 Robert Kiprono Cheruiyot  Kenya   2H 5M 52S  2.097778  Men's  
6  2008 Robert Kipkoech Cheruiyot Kenya   2H 7M 46S  2.129444  Men's  
7  2007 Robert Kipkoech Cheruiyot Kenya   2H 14M 13S 2.236944  Men's  
8  2006 Robert Kipkoech Cheruiyot Kenya   2H 7M 14S  2.120556  Men's  
9  2004 Timothy Cherigat          Kenya   2H 10M 37S 2.176944  Men's  
10 2003 Robert Kipkoech Cheruiyot Kenya   2H 10M 11S 2.169722  Men's  
11 2002 Rodgers Rop               Kenya   2H 9M 2S   2.150556  Men's  
12 2000 Elijah Lagat              Kenya   2H 9M 47S  2.163056  Men's  
13 1999 Joseph Chebet             Kenya   2H 9M 52S  2.164444  Men's  
14 1998 Moses Tanui               Kenya   2H 7M 34S  2.126111  Men's  
15 1997 Lameck Aguta              Kenya   2H 10M 34S 2.176111  Men's  
16 1996 Moses Tanui               Kenya   2H 9M 15S  2.154167  Men's  
17 1995 Cosmas Ndeti              Kenya   2H 9M 22S  2.156111  Men's  
18 1994 Cosmas Ndeti              Kenya   2H 7M 15S  2.120833  Men's  
19 1993 Cosmas Ndeti              Kenya   2H 9M 33S  2.159167  Men's  
20 1992 Ibrahim Hussein           Kenya   2H 8M 14S  2.137222  Men's  
21 1991 Ibrahim Hussein           Kenya   2H 11M 6S  2.185000  Men's  
22 1988 Ibrahim Hussein           Kenya   2H 8M 43S  2.145278  Men's  
23 2017 Edna Kiplagat             Kenya   2H 21M 52S 2.364444  Women's
24 2015 Caroline Rotich           Kenya   2H 24M 55S 2.415278  Women's
25 2013 Rita Jeptoo               Kenya   2H 26M 25S 2.440278  Women's
26 2012 Sharon Cherop             Kenya   2H 31M 50S 2.530556  Women's
27 2011 Caroline Kilel            Kenya   2H 22M 36S 2.376667  Women's
28 2009 Salina Kosgei             Kenya   2H 32M 16S 2.537778  Women's
29 2006 Rita Jeptoo               Kenya   2H 23M 38S 2.393889  Women's
30 2005 Catherine Ndereba         Kenya   2H 25M 13S 2.420278  Women's
31 2004 Catherine Ndereba         Kenya   2H 24M 27S 2.407500  Women's
32 2002 Margaret Okayo            Kenya   2H 20M 43S 2.345278  Women's
33 2001 Catherine Ndereba         Kenya   2H 23M 53S 2.398056  Women's
34 2000 Catherine Ndereba         Kenya   2H 26M 11S 2.436389  Women's
   country.simple
1  Kenya         
2  Kenya         
3  Kenya         
4  Kenya         
5  Kenya         
6  Kenya         
7  Kenya         
8  Kenya         
9  Kenya         
10 Kenya         
11 Kenya         
12 Kenya         
13 Kenya         
14 Kenya         
15 Kenya         
16 Kenya         
17 Kenya         
18 Kenya         
19 Kenya         
20 Kenya         
21 Kenya         
22 Kenya         
23 Kenya         
24 Kenya         
25 Kenya         
26 Kenya         
27 Kenya         
28 Kenya         
29 Kenya         
30 Kenya         
31 Kenya         
32 Kenya         
33 Kenya         
34 Kenya

# Select

- select let's you include/exclude columns.

In [4]:
# e.g.

# iris %>% select(Species, Sepal.Length) %>% head

Select just the columns that represent runner names and country from the marathon data.

In [24]:
df.marathon %>% 
    select(name, country) %>%
    head

name                       country      
1 Lawrence Cherono           Kenya        
2 Yuki Kawauchi              Japan        
3 Geoffrey Kirui             Kenya        
4 Lemi Berhanu               Ethiopia     
5 Lelisa Desisa              Ethiopia     
6 Mebrahtom "Meb" Keflezighi United States

# Putting it all together

Use `group_by`, `filter`, and `summarize` to show the best race times for US and Kenyan women runners. (*Hint: use the `%in%` operator to filter by two values for one column*). Fill in the missing pieces below:

In [31]:
df.marathon %>%
     filter(MF == "Women's", country %in% c("United States", "Kenya")) %>%
     group_by(country) %>%
     summarize(best.race.time = min(race.time))

country       best.race.time
1 Kenya         2.345278      
2 United States 2.378611

The best race time for US is 2.378611, for Kenya is 2.345278.

Use `filter`, `select`, `head`, and `arrange` to show the best women's race times in the data and the country of the runners. Fill in the missing pieces below:

In [34]:
df.marathon %>%
     filter(MF == "Women's") %>%
     select(race.time, country) %>%
     arrange(race.time)%>%
     head

race.time country      
1 2.333056  Ethiopia     
2 2.345278  Kenya        
3 2.362500  Germany      
4 2.364444  Kenya        
5 2.376667  Kenya        
6 2.378611  United States

The best woman's race time is 2.333056.

What's the best race time ever for a US man?

In [35]:
df.marathon %>%
    filter(MF == "Men's", country=="United States") %>%
    select(race.time, country) %>%
    arrange(race.time) %>%
    head

race.time country      
1 2.143611  United States
2 2.147778  United States
3 2.150000  United States
4 2.157500  United States
5 2.165278  United States
6 2.170278  United States

The best race time ever for a US man is 2.143611.

How does the best US men's time compare the the best race time in the data overall?

In [36]:
df.marathon %>%
    select(race.time, country) %>%
    arrange(race.time) %>%
    head

race.time country
1 2.050556  Kenya  
2 2.097778  Kenya  
3 2.120556  Kenya  
4 2.120833  Kenya  
5 2.126111  Kenya  
6 2.129444  Kenya

The best US men's time is 2.143611/2.050556 = 1.045 times as the best race time in the data overall.

# Bonus

For all countries with at least 5 winners in the men's category, which country has the best median race time in the men's category? (*Hint: use the `n()` function to count the winners by country.) 

In [30]:
# This should get you started. It adds a column "N.winners" that represents the
# total winners from the runner's country.

df.marathon %>%
    filter(MF == "Men's") %>%
    group_by(country) %>%
    mutate(N.winners = n()) %>% 
    filter(N.winners >=5) %>%
    summarize(median.racetime = median(race.time)) %>%
    arrange(desc(median.racetime))

country       median.racetime
1 Canada        2.503472       
2 United States 2.478889       
3 Finland       2.348333       
4 Japan         2.275833       
5 Ethiopia      2.163750       
6 Kenya         2.155139

Canada has the best median race time in the men's category.